{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create the files needed for projecting external data onto the [Stemformatics integrated atlases](stemformatics.org/atlas/myeloid). These files were used to generate the figures in the Stemformatics paper.\n",
    "\n",
    "Two external datasets were used: \n",
    "- Monkley et.al. RNAseq characterisation of iPSC-derived monocytes, macrophages and dendritic cells and their their primary blood derived counterparts (DOI: 10.1371/journal.pone.0243807)\n",
    "- Rosa et.al. Single-cell transcriptional profiling informs efficient reprogramming of human somatic cells to cross-presenting dendritic cells (DOI: 10.1126/sciimmunol.abg5539).\n",
    "\n",
    "Note that data files are not included in this repository due to size limits, but the providence of each input file used here \n",
    "is described below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set up the environment.\n",
    "import pandas, scanpy"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Process the Monkley data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CD14M1_02_S2</th>\n",
       "      <th>CD14M2_08_S8</th>\n",
       "      <th>CD14M3_16_S16</th>\n",
       "      <th>CD14P1_01_S1</th>\n",
       "      <th>CD14P2_07_S7</th>\n",
       "      <th>CD14P3_15_S15</th>\n",
       "      <th>IPSDM1M_rep1_03_S3</th>\n",
       "      <th>IPSDM1M_rep2_04_S4</th>\n",
       "      <th>IPSDM1P_rep1_05_S5</th>\n",
       "      <th>IPSDM1P_rep2_06_S6</th>\n",
       "      <th>...</th>\n",
       "      <th>iPSC_01_71_S29</th>\n",
       "      <th>iPSC_01_72_S30</th>\n",
       "      <th>iPSC_02_73_S31</th>\n",
       "      <th>iPSC_02_74_S32</th>\n",
       "      <th>iPSC_02_75_S33</th>\n",
       "      <th>iPSC_02_76_S34</th>\n",
       "      <th>iPSC_03_77_S35</th>\n",
       "      <th>iPSC_03_78_S36</th>\n",
       "      <th>iPSC_03_79_S37</th>\n",
       "      <th>iPSC_03_80_S38</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000000003</th>\n",
       "      <td>1.033403</td>\n",
       "      <td>1.681116</td>\n",
       "      <td>0.428623</td>\n",
       "      <td>0.902831</td>\n",
       "      <td>1.305014</td>\n",
       "      <td>0.361459</td>\n",
       "      <td>0.730497</td>\n",
       "      <td>1.013456</td>\n",
       "      <td>0.933722</td>\n",
       "      <td>0.772074</td>\n",
       "      <td>...</td>\n",
       "      <td>52.344791</td>\n",
       "      <td>62.641627</td>\n",
       "      <td>58.375401</td>\n",
       "      <td>55.022205</td>\n",
       "      <td>51.998891</td>\n",
       "      <td>52.198513</td>\n",
       "      <td>48.209343</td>\n",
       "      <td>50.188809</td>\n",
       "      <td>47.036905</td>\n",
       "      <td>46.441594</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000005</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.275243</td>\n",
       "      <td>0.272744</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.606626</td>\n",
       "      <td>0.548604</td>\n",
       "      <td>0.452728</td>\n",
       "      <td>0.209433</td>\n",
       "      <td>0.400881</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000419</th>\n",
       "      <td>26.504575</td>\n",
       "      <td>26.784796</td>\n",
       "      <td>29.651250</td>\n",
       "      <td>33.445910</td>\n",
       "      <td>41.452040</td>\n",
       "      <td>39.329630</td>\n",
       "      <td>38.077356</td>\n",
       "      <td>34.018670</td>\n",
       "      <td>38.700401</td>\n",
       "      <td>39.083730</td>\n",
       "      <td>...</td>\n",
       "      <td>34.079948</td>\n",
       "      <td>41.357203</td>\n",
       "      <td>36.759885</td>\n",
       "      <td>40.466421</td>\n",
       "      <td>37.733579</td>\n",
       "      <td>33.718873</td>\n",
       "      <td>41.363873</td>\n",
       "      <td>35.603303</td>\n",
       "      <td>41.113588</td>\n",
       "      <td>38.529955</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000457</th>\n",
       "      <td>3.357820</td>\n",
       "      <td>6.799516</td>\n",
       "      <td>4.963140</td>\n",
       "      <td>5.851235</td>\n",
       "      <td>5.846110</td>\n",
       "      <td>4.778542</td>\n",
       "      <td>3.818070</td>\n",
       "      <td>4.146496</td>\n",
       "      <td>1.945063</td>\n",
       "      <td>2.024698</td>\n",
       "      <td>...</td>\n",
       "      <td>2.510417</td>\n",
       "      <td>2.627912</td>\n",
       "      <td>3.379397</td>\n",
       "      <td>3.547168</td>\n",
       "      <td>3.736435</td>\n",
       "      <td>3.806338</td>\n",
       "      <td>2.156203</td>\n",
       "      <td>2.529889</td>\n",
       "      <td>2.676792</td>\n",
       "      <td>2.881431</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000460</th>\n",
       "      <td>3.398016</td>\n",
       "      <td>6.340220</td>\n",
       "      <td>4.897647</td>\n",
       "      <td>9.083736</td>\n",
       "      <td>11.662446</td>\n",
       "      <td>10.029129</td>\n",
       "      <td>8.221695</td>\n",
       "      <td>7.105386</td>\n",
       "      <td>7.630245</td>\n",
       "      <td>5.857991</td>\n",
       "      <td>...</td>\n",
       "      <td>15.364644</td>\n",
       "      <td>22.091836</td>\n",
       "      <td>18.997053</td>\n",
       "      <td>22.503723</td>\n",
       "      <td>26.353436</td>\n",
       "      <td>20.344215</td>\n",
       "      <td>15.104561</td>\n",
       "      <td>16.562847</td>\n",
       "      <td>13.657067</td>\n",
       "      <td>13.447602</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 98 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 CD14M1_02_S2  CD14M2_08_S8  CD14M3_16_S16  CD14P1_01_S1  \\\n",
       "gene_id                                                                    \n",
       "ENSG00000000003      1.033403      1.681116       0.428623      0.902831   \n",
       "ENSG00000000005      0.000000      0.000000       0.000000      0.000000   \n",
       "ENSG00000000419     26.504575     26.784796      29.651250     33.445910   \n",
       "ENSG00000000457      3.357820      6.799516       4.963140      5.851235   \n",
       "ENSG00000000460      3.398016      6.340220       4.897647      9.083736   \n",
       "\n",
       "                 CD14P2_07_S7  CD14P3_15_S15  IPSDM1M_rep1_03_S3  \\\n",
       "gene_id                                                            \n",
       "ENSG00000000003      1.305014       0.361459            0.730497   \n",
       "ENSG00000000005      0.000000       0.000000            0.000000   \n",
       "ENSG00000000419     41.452040      39.329630           38.077356   \n",
       "ENSG00000000457      5.846110       4.778542            3.818070   \n",
       "ENSG00000000460     11.662446      10.029129            8.221695   \n",
       "\n",
       "                 IPSDM1M_rep2_04_S4  IPSDM1P_rep1_05_S5  IPSDM1P_rep2_06_S6  \\\n",
       "gene_id                                                                       \n",
       "ENSG00000000003            1.013456            0.933722            0.772074   \n",
       "ENSG00000000005            0.000000            0.000000            0.000000   \n",
       "ENSG00000000419           34.018670           38.700401           39.083730   \n",
       "ENSG00000000457            4.146496            1.945063            2.024698   \n",
       "ENSG00000000460            7.105386            7.630245            5.857991   \n",
       "\n",
       "                 ...  iPSC_01_71_S29  iPSC_01_72_S30  iPSC_02_73_S31  \\\n",
       "gene_id          ...                                                   \n",
       "ENSG00000000003  ...       52.344791       62.641627       58.375401   \n",
       "ENSG00000000005  ...        0.275243        0.272744        0.000000   \n",
       "ENSG00000000419  ...       34.079948       41.357203       36.759885   \n",
       "ENSG00000000457  ...        2.510417        2.627912        3.379397   \n",
       "ENSG00000000460  ...       15.364644       22.091836       18.997053   \n",
       "\n",
       "                 iPSC_02_74_S32  iPSC_02_75_S33  iPSC_02_76_S34  \\\n",
       "gene_id                                                           \n",
       "ENSG00000000003       55.022205       51.998891       52.198513   \n",
       "ENSG00000000005        0.606626        0.548604        0.452728   \n",
       "ENSG00000000419       40.466421       37.733579       33.718873   \n",
       "ENSG00000000457        3.547168        3.736435        3.806338   \n",
       "ENSG00000000460       22.503723       26.353436       20.344215   \n",
       "\n",
       "                 iPSC_03_77_S35  iPSC_03_78_S36  iPSC_03_79_S37  \\\n",
       "gene_id                                                           \n",
       "ENSG00000000003       48.209343       50.188809       47.036905   \n",
       "ENSG00000000005        0.209433        0.400881        0.000000   \n",
       "ENSG00000000419       41.363873       35.603303       41.113588   \n",
       "ENSG00000000457        2.156203        2.529889        2.676792   \n",
       "ENSG00000000460       15.104561       16.562847       13.657067   \n",
       "\n",
       "                 iPSC_03_80_S38  \n",
       "gene_id                          \n",
       "ENSG00000000003       46.441594  \n",
       "ENSG00000000005        0.000000  \n",
       "ENSG00000000419       38.529955  \n",
       "ENSG00000000457        2.881431  \n",
       "ENSG00000000460       13.447602  \n",
       "\n",
       "[5 rows x 98 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Comment[ENA_SAMPLE]</th>\n",
       "      <th>Comment[BioSD_SAMPLE]</th>\n",
       "      <th>Characteristics[organism]</th>\n",
       "      <th>Characteristics[organism part]</th>\n",
       "      <th>Characteristics[cell line]</th>\n",
       "      <th>Characteristics[cell type]</th>\n",
       "      <th>Characteristics[donor]</th>\n",
       "      <th>Characteristics[batch]</th>\n",
       "      <th>Characteristics[run]</th>\n",
       "      <th>Characteristics[harvest]</th>\n",
       "      <th>...</th>\n",
       "      <th>Performer.5</th>\n",
       "      <th>Protocol REF.6</th>\n",
       "      <th>Performer.6</th>\n",
       "      <th>Derived Array Data File</th>\n",
       "      <th>Protocol REF.7</th>\n",
       "      <th>Performer.7</th>\n",
       "      <th>Protocol REF.8</th>\n",
       "      <th>Performer.8</th>\n",
       "      <th>Derived Array Data File.1</th>\n",
       "      <th>Factor Value[cell type]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Source Name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CD14M1_02_S2</th>\n",
       "      <td>ERS5183935</td>\n",
       "      <td>SAMEA7425716</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R-iPS 1J</td>\n",
       "      <td>induced pluripotent stem cell</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>d24</td>\n",
       "      <td>...</td>\n",
       "      <td>AstraZeneca Gothenburg RIA Bioinformatics</td>\n",
       "      <td>P-MTAB-102037</td>\n",
       "      <td>AstraZeneca Gothenburg RIA Bioinformatics</td>\n",
       "      <td>iPSd_comb_counts.txt</td>\n",
       "      <td>P-MTAB-102040</td>\n",
       "      <td>AstraZeneca Gothenburg RIA Bioinformatics</td>\n",
       "      <td>P-MTAB-102037</td>\n",
       "      <td>AstraZeneca Gothenburg RIA Bioinformatics</td>\n",
       "      <td>iPSCd_comb_geneTPM.txt</td>\n",
       "      <td>induced pluripotent stem cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CD14M1_02_S2</th>\n",
       "      <td>ERS5183935</td>\n",
       "      <td>SAMEA7425716</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R-iPS 1J</td>\n",
       "      <td>induced pluripotent stem cell</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>d24</td>\n",
       "      <td>...</td>\n",
       "      <td>AstraZeneca Gothenburg RIA Bioinformatics</td>\n",
       "      <td>P-MTAB-102037</td>\n",
       "      <td>AstraZeneca Gothenburg RIA Bioinformatics</td>\n",
       "      <td>iPSd_comb_counts.txt</td>\n",
       "      <td>P-MTAB-102040</td>\n",
       "      <td>AstraZeneca Gothenburg RIA Bioinformatics</td>\n",
       "      <td>P-MTAB-102037</td>\n",
       "      <td>AstraZeneca Gothenburg RIA Bioinformatics</td>\n",
       "      <td>iPSCd_comb_geneTPM.txt</td>\n",
       "      <td>induced pluripotent stem cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CD14M2_08_S8</th>\n",
       "      <td>ERS5183936</td>\n",
       "      <td>SAMEA7425717</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R-iPS 1J</td>\n",
       "      <td>induced pluripotent stem cell</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>d31</td>\n",
       "      <td>...</td>\n",
       "      <td>AstraZeneca Gothenburg RIA Bioinformatics</td>\n",
       "      <td>P-MTAB-102037</td>\n",
       "      <td>AstraZeneca Gothenburg RIA Bioinformatics</td>\n",
       "      <td>iPSd_comb_counts.txt</td>\n",
       "      <td>P-MTAB-102040</td>\n",
       "      <td>AstraZeneca Gothenburg RIA Bioinformatics</td>\n",
       "      <td>P-MTAB-102037</td>\n",
       "      <td>AstraZeneca Gothenburg RIA Bioinformatics</td>\n",
       "      <td>iPSCd_comb_geneTPM.txt</td>\n",
       "      <td>induced pluripotent stem cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CD14M2_08_S8</th>\n",
       "      <td>ERS5183936</td>\n",
       "      <td>SAMEA7425717</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R-iPS 1J</td>\n",
       "      <td>induced pluripotent stem cell</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>d31</td>\n",
       "      <td>...</td>\n",
       "      <td>AstraZeneca Gothenburg RIA Bioinformatics</td>\n",
       "      <td>P-MTAB-102037</td>\n",
       "      <td>AstraZeneca Gothenburg RIA Bioinformatics</td>\n",
       "      <td>iPSd_comb_counts.txt</td>\n",
       "      <td>P-MTAB-102040</td>\n",
       "      <td>AstraZeneca Gothenburg RIA Bioinformatics</td>\n",
       "      <td>P-MTAB-102037</td>\n",
       "      <td>AstraZeneca Gothenburg RIA Bioinformatics</td>\n",
       "      <td>iPSCd_comb_geneTPM.txt</td>\n",
       "      <td>induced pluripotent stem cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CD14M3_16_S16</th>\n",
       "      <td>ERS5183937</td>\n",
       "      <td>SAMEA7425718</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R-iPS 1J</td>\n",
       "      <td>induced pluripotent stem cell</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>d38</td>\n",
       "      <td>...</td>\n",
       "      <td>AstraZeneca Gothenburg RIA Bioinformatics</td>\n",
       "      <td>P-MTAB-102037</td>\n",
       "      <td>AstraZeneca Gothenburg RIA Bioinformatics</td>\n",
       "      <td>iPSd_comb_counts.txt</td>\n",
       "      <td>P-MTAB-102040</td>\n",
       "      <td>AstraZeneca Gothenburg RIA Bioinformatics</td>\n",
       "      <td>P-MTAB-102037</td>\n",
       "      <td>AstraZeneca Gothenburg RIA Bioinformatics</td>\n",
       "      <td>iPSCd_comb_geneTPM.txt</td>\n",
       "      <td>induced pluripotent stem cell</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 51 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              Comment[ENA_SAMPLE] Comment[BioSD_SAMPLE]  \\\n",
       "Source Name                                               \n",
       "CD14M1_02_S2           ERS5183935          SAMEA7425716   \n",
       "CD14M1_02_S2           ERS5183935          SAMEA7425716   \n",
       "CD14M2_08_S8           ERS5183936          SAMEA7425717   \n",
       "CD14M2_08_S8           ERS5183936          SAMEA7425717   \n",
       "CD14M3_16_S16          ERS5183937          SAMEA7425718   \n",
       "\n",
       "              Characteristics[organism] Characteristics[organism part]  \\\n",
       "Source Name                                                              \n",
       "CD14M1_02_S2               Homo sapiens                            NaN   \n",
       "CD14M1_02_S2               Homo sapiens                            NaN   \n",
       "CD14M2_08_S8               Homo sapiens                            NaN   \n",
       "CD14M2_08_S8               Homo sapiens                            NaN   \n",
       "CD14M3_16_S16              Homo sapiens                            NaN   \n",
       "\n",
       "              Characteristics[cell line]     Characteristics[cell type]  \\\n",
       "Source Name                                                               \n",
       "CD14M1_02_S2                    R-iPS 1J  induced pluripotent stem cell   \n",
       "CD14M1_02_S2                    R-iPS 1J  induced pluripotent stem cell   \n",
       "CD14M2_08_S8                    R-iPS 1J  induced pluripotent stem cell   \n",
       "CD14M2_08_S8                    R-iPS 1J  induced pluripotent stem cell   \n",
       "CD14M3_16_S16                   R-iPS 1J  induced pluripotent stem cell   \n",
       "\n",
       "               Characteristics[donor]  Characteristics[batch]  \\\n",
       "Source Name                                                     \n",
       "CD14M1_02_S2                        0                       1   \n",
       "CD14M1_02_S2                        0                       1   \n",
       "CD14M2_08_S8                        0                       1   \n",
       "CD14M2_08_S8                        0                       1   \n",
       "CD14M3_16_S16                       0                       1   \n",
       "\n",
       "               Characteristics[run] Characteristics[harvest]  ...  \\\n",
       "Source Name                                                   ...   \n",
       "CD14M1_02_S2                      1                      d24  ...   \n",
       "CD14M1_02_S2                      1                      d24  ...   \n",
       "CD14M2_08_S8                      1                      d31  ...   \n",
       "CD14M2_08_S8                      1                      d31  ...   \n",
       "CD14M3_16_S16                     1                      d38  ...   \n",
       "\n",
       "                                             Performer.5 Protocol REF.6  \\\n",
       "Source Name                                                               \n",
       "CD14M1_02_S2   AstraZeneca Gothenburg RIA Bioinformatics  P-MTAB-102037   \n",
       "CD14M1_02_S2   AstraZeneca Gothenburg RIA Bioinformatics  P-MTAB-102037   \n",
       "CD14M2_08_S8   AstraZeneca Gothenburg RIA Bioinformatics  P-MTAB-102037   \n",
       "CD14M2_08_S8   AstraZeneca Gothenburg RIA Bioinformatics  P-MTAB-102037   \n",
       "CD14M3_16_S16  AstraZeneca Gothenburg RIA Bioinformatics  P-MTAB-102037   \n",
       "\n",
       "                                             Performer.6  \\\n",
       "Source Name                                                \n",
       "CD14M1_02_S2   AstraZeneca Gothenburg RIA Bioinformatics   \n",
       "CD14M1_02_S2   AstraZeneca Gothenburg RIA Bioinformatics   \n",
       "CD14M2_08_S8   AstraZeneca Gothenburg RIA Bioinformatics   \n",
       "CD14M2_08_S8   AstraZeneca Gothenburg RIA Bioinformatics   \n",
       "CD14M3_16_S16  AstraZeneca Gothenburg RIA Bioinformatics   \n",
       "\n",
       "              Derived Array Data File Protocol REF.7  \\\n",
       "Source Name                                            \n",
       "CD14M1_02_S2     iPSd_comb_counts.txt  P-MTAB-102040   \n",
       "CD14M1_02_S2     iPSd_comb_counts.txt  P-MTAB-102040   \n",
       "CD14M2_08_S8     iPSd_comb_counts.txt  P-MTAB-102040   \n",
       "CD14M2_08_S8     iPSd_comb_counts.txt  P-MTAB-102040   \n",
       "CD14M3_16_S16    iPSd_comb_counts.txt  P-MTAB-102040   \n",
       "\n",
       "                                             Performer.7 Protocol REF.8  \\\n",
       "Source Name                                                               \n",
       "CD14M1_02_S2   AstraZeneca Gothenburg RIA Bioinformatics  P-MTAB-102037   \n",
       "CD14M1_02_S2   AstraZeneca Gothenburg RIA Bioinformatics  P-MTAB-102037   \n",
       "CD14M2_08_S8   AstraZeneca Gothenburg RIA Bioinformatics  P-MTAB-102037   \n",
       "CD14M2_08_S8   AstraZeneca Gothenburg RIA Bioinformatics  P-MTAB-102037   \n",
       "CD14M3_16_S16  AstraZeneca Gothenburg RIA Bioinformatics  P-MTAB-102037   \n",
       "\n",
       "                                             Performer.8  \\\n",
       "Source Name                                                \n",
       "CD14M1_02_S2   AstraZeneca Gothenburg RIA Bioinformatics   \n",
       "CD14M1_02_S2   AstraZeneca Gothenburg RIA Bioinformatics   \n",
       "CD14M2_08_S8   AstraZeneca Gothenburg RIA Bioinformatics   \n",
       "CD14M2_08_S8   AstraZeneca Gothenburg RIA Bioinformatics   \n",
       "CD14M3_16_S16  AstraZeneca Gothenburg RIA Bioinformatics   \n",
       "\n",
       "              Derived Array Data File.1        Factor Value[cell type]  \n",
       "Source Name                                                             \n",
       "CD14M1_02_S2     iPSCd_comb_geneTPM.txt  induced pluripotent stem cell  \n",
       "CD14M1_02_S2     iPSCd_comb_geneTPM.txt  induced pluripotent stem cell  \n",
       "CD14M2_08_S8     iPSCd_comb_geneTPM.txt  induced pluripotent stem cell  \n",
       "CD14M2_08_S8     iPSCd_comb_geneTPM.txt  induced pluripotent stem cell  \n",
       "CD14M3_16_S16    iPSCd_comb_geneTPM.txt  induced pluripotent stem cell  \n",
       "\n",
       "[5 rows x 51 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shape of expression matrix: (56226, 98) Shape of sample matrix: (178, 51)\n",
      "['Comment[ENA_SAMPLE]', 'Comment[BioSD_SAMPLE]', 'Characteristics[organism]', 'Characteristics[organism part]', 'Characteristics[cell line]', 'Characteristics[cell type]', 'Characteristics[donor]', 'Characteristics[batch]', 'Characteristics[run]', 'Characteristics[harvest]', 'Characteristics[stimulus]', 'Characteristics[stimulus_time]', 'Characteristics[cell_type_abbrev]', 'Material Type', 'Description', 'Protocol REF', 'Performer', 'Protocol REF.1', 'Performer.1', 'Protocol REF.2', 'Performer.2', 'Protocol REF.3', 'Performer.3', 'Extract Name', 'Comment[LIBRARY_LAYOUT]', 'Comment[LIBRARY_SELECTION]', 'Comment[LIBRARY_SOURCE]', 'Comment[LIBRARY_STRAND]', 'Comment[LIBRARY_STRATEGY]', 'Comment[NOMINAL_LENGTH]', 'Comment[NOMINAL_SDEV]', 'Protocol REF.4', 'Performer.4', 'Assay Name', 'Technology Type', 'Comment[ENA_EXPERIMENT]', 'Scan Name', 'Comment[SUBMITTED_FILE_NAME]', 'Comment[ENA_RUN]', 'Comment[FASTQ_URI]', 'Protocol REF.5', 'Performer.5', 'Protocol REF.6', 'Performer.6', 'Derived Array Data File', 'Protocol REF.7', 'Performer.7', 'Protocol REF.8', 'Performer.8', 'Derived Array Data File.1', 'Factor Value[cell type]']\n",
      "Shape of expression matrix: (56226, 97) Shape of sample matrix: (97, 5)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>combined</th>\n",
       "      <th>cell_type_abbrev</th>\n",
       "      <th>cell type</th>\n",
       "      <th>cell line</th>\n",
       "      <th>stimulus</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>IPSDDC_2D_d32_30_S28</th>\n",
       "      <td>IPSCDC</td>\n",
       "      <td>IPSCDC</td>\n",
       "      <td>induced pluripotent stem cell</td>\n",
       "      <td>R-iPS 1J</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IPSDM_2B_d21_6h_19_S17</th>\n",
       "      <td>IPSDM_LPS</td>\n",
       "      <td>IPSDM</td>\n",
       "      <td>induced pluripotent stem cell</td>\n",
       "      <td>R-iPS 1J</td>\n",
       "      <td>LPS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CD14M2_08_S8</th>\n",
       "      <td>IPSDMo_CD14pos</td>\n",
       "      <td>IPSDMo_CD14pos</td>\n",
       "      <td>induced pluripotent stem cell</td>\n",
       "      <td>R-iPS 1J</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>iPSC_02_73_S31</th>\n",
       "      <td>IPSC</td>\n",
       "      <td>IPSC</td>\n",
       "      <td>induced pluripotent stem cell</td>\n",
       "      <td>R-iPS 1J</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HMDDC_D2_24h_54_S12</th>\n",
       "      <td>MoDC_LPS</td>\n",
       "      <td>MoDC</td>\n",
       "      <td>blood cell</td>\n",
       "      <td></td>\n",
       "      <td>LPS</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              combined cell_type_abbrev  \\\n",
       "IPSDDC_2D_d32_30_S28            IPSCDC           IPSCDC   \n",
       "IPSDM_2B_d21_6h_19_S17       IPSDM_LPS            IPSDM   \n",
       "CD14M2_08_S8            IPSDMo_CD14pos   IPSDMo_CD14pos   \n",
       "iPSC_02_73_S31                    IPSC             IPSC   \n",
       "HMDDC_D2_24h_54_S12           MoDC_LPS             MoDC   \n",
       "\n",
       "                                            cell type cell line stimulus  \n",
       "IPSDDC_2D_d32_30_S28    induced pluripotent stem cell  R-iPS 1J     none  \n",
       "IPSDM_2B_d21_6h_19_S17  induced pluripotent stem cell  R-iPS 1J      LPS  \n",
       "CD14M2_08_S8            induced pluripotent stem cell  R-iPS 1J     none  \n",
       "iPSC_02_73_S31          induced pluripotent stem cell  R-iPS 1J     none  \n",
       "HMDDC_D2_24h_54_S12                        blood cell                LPS  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "\"\"\"Process the Monkley data. The original gene expression matrix and sample metadata were downloaded from ArrayExpress:\n",
    "https://www.ebi.ac.uk/biostudies/arrayexpress/studies/E-MTAB-9670. \n",
    "Gene expression file: iPSCd_comb_geneTPM.txt\n",
    "Sample metadata file: E-MTAB-9670.sdrf.txt\n",
    "All other files are processed and produced here.\n",
    "\"\"\"\n",
    "def processMonkley():\n",
    "    # Show what original data look like\n",
    "    df = pandas.read_csv(\"data/Monkley/iPSCd_comb_geneTPM.txt\", sep=\"\\t\", index_col=0)\n",
    "    samples = pandas.read_csv(\"data/Monkley/E-MTAB-9670.sdrf.txt\", sep=\"\\t\", index_col=0)\n",
    "    display(df.head())\n",
    "    display(samples.head())\n",
    "    print(\"Shape of expression matrix:\", df.shape, \"Shape of sample matrix:\", samples.shape)\n",
    "\n",
    "    # Sample matrix has more samples than there are in the expression matrix.\n",
    "    # Create files where the samples are matched between the two.\n",
    "    \n",
    "    commonSampleIds = set(df.columns).intersection(set(samples.index))\n",
    "    df = df[list(commonSampleIds)]\n",
    "    df.to_csv(\"data/Monkley/iPSCd_comb_geneTPM.SampleMatched.tsv\", sep=\"\\t\")\n",
    "    \n",
    "    # We also don't need all the columns in the sample matrix - select a few we want and rename them too.\n",
    "    print(samples.columns.tolist())\n",
    "    samples = samples[['Characteristics[cell_type_abbrev]','Characteristics[cell type]','Characteristics[cell line]','Characteristics[stimulus]']]\n",
    "    samples = samples.loc[df.columns].fillna('')\n",
    "    samples = samples[~samples.index.duplicated(keep='first')]\n",
    "    samples.columns = [item.replace('Characteristics[','').replace(']','') for item in samples.columns]\n",
    "    \n",
    "    # Insert a new column in samples called \"combined\" which combines cell type with stimulus\n",
    "    samples.insert(0, 'combined', [f\"{row['cell_type_abbrev']}_{row['stimulus']}\".replace(\"_none\",\"\") for index,row in samples.iterrows()])\n",
    "    \n",
    "    print(\"Shape of expression matrix:\", df.shape, \"Shape of sample matrix:\", samples.shape)\n",
    "    display(samples.head())\n",
    "    samples.to_csv(\"data/Monkley/samples.tsv\", sep=\"\\t\")\n",
    "    \n",
    "processMonkley()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Do specified sample ids all belong to the sample matrix? True\n",
      "Filtered sample matrix shape: (54, 5)\n"
     ]
    }
   ],
   "source": [
    "\"\"\"For the purpose of creating figures for the paper, we further filtered out some samples for simplicity, \n",
    "as it can be hard to see patterns with too many cell types on a static printed figure (it works much better\n",
    "as an interactive plot on the website when above files are used directly to project the data).\n",
    "\"\"\"\n",
    "def createMonkleySubset():\n",
    "    samples = pandas.read_csv(\"data/Monkley/samples.tsv\", sep=\"\\t\", index_col=0)\n",
    "    keep = ['HMDM_D3_unstim_63_S21', 'HMDM_D2_unstim_60_S18', 'HMDM_D1_unstim_57_S15', 'IPSDDC_1D_d21_P3_9_S7', \n",
    "            'IPSDDC_2D_d32_30_S28', 'IPSDDC_2D_d22_28_S26', 'IPSDDC_3D_d32_47_S7', 'IPSDDC_2D_d29_29_S27', \n",
    "            'IPSDDC_1D_d35_P4_13_S11', 'IPSDDC_1D_d28_P3_11_S9', 'IPSDDC_3D_d22_45_S5', 'IPSDDC_3D_d29_46_S6', \n",
    "            'IPSDM1M_rep1_03_S3', 'IPSDM_3B_d21_unstim_34_S32', 'IPSDM3M_rep2_18_S18', 'IPSDM3M_rep3_19_S19', \n",
    "            'IPSDM_2B_d28_unstim_20_S18', 'IPSDM2M_rep3_11_S11', 'IPSDM_3B_d35_unstim_40_S38', 'IPSDM2M_rep1_09_S9', \n",
    "            'IPSDM_2B_d21_unstim_17_S15', 'IPSDM1M_rep2_04_S4', 'IPSDM_2B_d35_unstim_23_S21', \n",
    "            'IPSDM_3B_d28_unstim_37_S35', 'IPSDM3M_rep1_17_S17', 'IPSDM2M_rep2_10_S10', 'CD14N_1D_d35_4_S3', \n",
    "            'CD14N_1D_d31_3_S2', 'CD14N_1D_d28_2_S1', 'CD14N_2D_d43_26_S24', 'CD14P2_07_S7', 'CD14P3_15_S15', \n",
    "            'CD14N_3D_d43_43_S3', 'CD14P1_01_S1', 'CD14P_3B_d21_31_S29', 'CD14P_3D_d43_44_S4', 'CD14M2_08_S8', \n",
    "            'CD14P_2B_d21_14_S12', 'CD14P_3B_d28_32_S30', 'CD14P_1D_d31_7_S5', 'CD14P_2B_d35_16_S14', \n",
    "            'CD14M3_16_S16', 'CD14P_1D_d35_8_S6', 'CD14P_1D_d28_6_S4', 'CD14P_3B_d35_33_S31', 'CD14P_2B_d28_15_S13', \n",
    "            'CD14P_2D_d43_27_S25', 'CD14M1_02_S2', 'HMDDC_D2_unstim_53_S11', 'HMDDC_D3_unstim_55_S13', \n",
    "            'HMDDC_D1_unstim_51_S9', 'PBMC_D3_68_S26', 'PBMC_D1_66_S24', 'PBMC_D2_67_S25']\n",
    "    print(\"Do specified sample ids all belong to the sample matrix?\", set(keep).issubset(set(samples.index)))\n",
    "    # - in case there was a mistake in this step\n",
    "    \n",
    "    samples = samples.loc[keep]\n",
    "    print(\"Filtered sample matrix shape:\", samples.shape)\n",
    "    \n",
    "    # Write the expression and sample matrices which match these samples to files.\n",
    "    # These were the files used to make projections onto the Stemformatics dendritic cell atlas \n",
    "    # (stemformatics.org/atlas/dc). Screenshots were then used to create the figures for the paper.\n",
    "    df = pandas.read_csv(\"data/Monkley/iPSCd_comb_geneTPM.SampleMatched.tsv\", sep=\"\\t\", index_col=0)\n",
    "    df = df[samples.index]\n",
    "    df.to_csv(\"data/Monkley/iPSCd_comb_geneTPM.SampleMatched.subset.tsv\", sep=\"\\t\")\n",
    "    samples.to_csv(\"data/Monkley/samples.subset.tsv\", sep=\"\\t\")\n",
    "    \n",
    "createMonkleySubset()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Process the Rosa data\n",
    "We started by downloading GSE162650_hef_all_counts.txt from [GEO](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE162650). Then we ran the following R code to convert this to a sparse matrix, matrix.mtx.\n",
    "Other files created by this script include genes.tsv (gene symbols corresponding to the features of the matrix),\n",
    "barcodes.tsv (barcodes corresponding to the samples of the matrix) and celltypes.tsv (cell type that each sample\n",
    "belongs to).\n",
    "\n",
    "```R\n",
    ".libPaths('~/libs/R_libs')\n",
    "require(data.table)\n",
    "require(Matrix)\n",
    "dat <- fread('./GSE162650_hef_all_counts.txt')\n",
    "barcode <- colnames(dat)[-1]\n",
    "celltype <- substr(barcode,1,nchar(barcode)-19)\n",
    "unique_celltype <- unique(celltype)\n",
    "gene <- as.vector(as.matrix(dat[,1]))\n",
    "dat <- dat[,-1]\n",
    "dat <- Matrix(as.matrix(data.frame(dat)),sparse = TRUE)\n",
    "writeMM(obj = dat, file = paste(\".\",\"/matrix.mtx\",sep = ''))\n",
    "write(x = gene, file = paste(\".\",\"/genes.tsv\",sep = ''))\n",
    "write(x = barcode, file = paste(\".\",\"/barcodes.tsv\",sep = ''))\n",
    "write(x = celltype, file = paste(\".\",\"/celltypes.tsv\",sep = ''))\n",
    "```\n",
    "\n",
    "We then used scanpy's read_mtx function to read matrix.mtx, and assigned the genes from genes.tsv as its features (vars) and barcodes from barcodes.tsv as its samples (obs). We also log normalised using scanpy's log1p function (to base 2). \n",
    "This created LogNormalised.h5ad file.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"Function to output a pseudo-bulk aggregated data, given single cell data as input. \n",
    "> pseudoBulkAggregate(df, ['HSC','B','B',...])\n",
    "\n",
    "df (pandas.DataFrame): genes (rows) x samples (columns) expression matrix (single cell RNA-seq dataset)\n",
    "groups (pandas.Series): specifies what group each sample belongs to (eg. cell type); index equals df.columns\n",
    "n (int): how many points to represent the smallest cluster with\n",
    "\n",
    "This function first determines the smallest group based on groups, then tries to split that group into n sub-groups.\n",
    "This determines the cluster size which will be used to aggregate. So if the smallest group has 30 samples in it,\n",
    "for example, and n=3, 10 samples are used to create each aggregated sample for each group. Hence keep an eye on\n",
    "the sample size, as it may not work well if this is too small.\n",
    "\"\"\"\n",
    "def pseudoBulkAggregate(df, groups, n=3, verbose=True):\n",
    "    import random\n",
    "    \n",
    "    valueCounts = groups.value_counts().sort_values()\n",
    "    smallest = valueCounts.tolist()[0] # size of the smallest cluster\n",
    "    if verbose: print(\"value counts of cells in the group\", valueCounts)\n",
    "    \n",
    "    if smallest<=3*n: # if the smallest cluster isn't at least 3 times the value of n, just use all of that cluster\n",
    "        n = 1\n",
    "    sampleSize = int(smallest/n)  # how many samples to put into each aggregated cluster\n",
    "    if sampleSize==1:  # no aggregation happening\n",
    "        raise Exception(\"Could not determine how many samples to aggregate together. Likely due to a cluster having only one member.\")\n",
    "    if verbose: print(\"sample size for each cluster:\", sampleSize)\n",
    "\n",
    "    aggregated = {}\n",
    "    for item in valueCounts.index:  # each item of group\n",
    "        subset = df[groups[groups==item].index]  # expression for this sampleGroup\n",
    "        # we need new sample ids based on sample group eg. \"Mono1__0\"\n",
    "        i = 0\n",
    "        columns = subset.columns.tolist()\n",
    "        while len(columns)>sampleSize:\n",
    "            selectedColumns = random.sample(list(columns), sampleSize)\n",
    "            aggregated[\"%s__%s\" % (item, i)] = subset[selectedColumns].sum(axis=1)\n",
    "            columns = set(columns).difference(set(selectedColumns))\n",
    "            i += 1\n",
    "\n",
    "    return pandas.DataFrame(aggregated, index=df.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "value counts of cells in the group Day9_SP_B       73\n",
      "DC1_A          113\n",
      "DC1_B          160\n",
      "PDC_A          210\n",
      "Day9_SP_C      228\n",
      "DC1_C          257\n",
      "HEF_C          275\n",
      "Day9_DP_A1     280\n",
      "Day3_A1        403\n",
      "Day9_SP_A      411\n",
      "HEF_B          578\n",
      "DC2_A          615\n",
      "Day9_DP_C      631\n",
      "HEF_A1         812\n",
      "Day9_DP_A2     861\n",
      "PDC_B          964\n",
      "Day3_B        1028\n",
      "Day6_A        1031\n",
      "Day9_DP_B     1076\n",
      "PDC_C         1115\n",
      "DC2_B         1872\n",
      "Day3_A2       1907\n",
      "DC2_C         2067\n",
      "HEF_A2        2411\n",
      "Day3_C        3292\n",
      "Day6_B        3811\n",
      "Day6_C        7044\n",
      "dtype: int64\n",
      "sample size for each cluster: 24\n",
      "(29880, 1382)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Day9_SP_B__0</th>\n",
       "      <th>Day9_SP_B__1</th>\n",
       "      <th>Day9_SP_B__2</th>\n",
       "      <th>DC1_A__0</th>\n",
       "      <th>DC1_A__1</th>\n",
       "      <th>DC1_A__2</th>\n",
       "      <th>DC1_A__3</th>\n",
       "      <th>DC1_B__0</th>\n",
       "      <th>DC1_B__1</th>\n",
       "      <th>DC1_B__2</th>\n",
       "      <th>...</th>\n",
       "      <th>Day6_C__283</th>\n",
       "      <th>Day6_C__284</th>\n",
       "      <th>Day6_C__285</th>\n",
       "      <th>Day6_C__286</th>\n",
       "      <th>Day6_C__287</th>\n",
       "      <th>Day6_C__288</th>\n",
       "      <th>Day6_C__289</th>\n",
       "      <th>Day6_C__290</th>\n",
       "      <th>Day6_C__291</th>\n",
       "      <th>Day6_C__292</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AL627309.1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.133147</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL627309.5</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.407509</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LINC01409</th>\n",
       "      <td>0.821351</td>\n",
       "      <td>0.849008</td>\n",
       "      <td>1.061065</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.431754</td>\n",
       "      <td>4.811491</td>\n",
       "      <td>6.380960</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.236782</td>\n",
       "      <td>0.098364</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LINC01128</th>\n",
       "      <td>0.775091</td>\n",
       "      <td>1.342818</td>\n",
       "      <td>1.813515</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.295059</td>\n",
       "      <td>3.752775</td>\n",
       "      <td>4.724329</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.514160</td>\n",
       "      <td>0.187913</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.098073</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LINC00115</th>\n",
       "      <td>0.616771</td>\n",
       "      <td>0.362355</td>\n",
       "      <td>0.221942</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.550306</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.053464</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.446060</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.050391</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.084598</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.098105</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 1382 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Day9_SP_B__0  Day9_SP_B__1  Day9_SP_B__2  DC1_A__0  DC1_A__1  \\\n",
       "AL627309.1      0.000000      0.000000      0.133147       0.0       0.0   \n",
       "AL627309.5      0.000000      0.000000      0.000000       0.0       0.0   \n",
       "LINC01409       0.821351      0.849008      1.061065       0.0       0.0   \n",
       "LINC01128       0.775091      1.342818      1.813515       0.0       0.0   \n",
       "LINC00115       0.616771      0.362355      0.221942       0.0       0.0   \n",
       "\n",
       "            DC1_A__2  DC1_A__3  DC1_B__0  DC1_B__1  DC1_B__2  ...  \\\n",
       "AL627309.1  0.000000       0.0  0.000000  0.000000  0.000000  ...   \n",
       "AL627309.5  0.000000       0.0  0.000000  0.407509  0.000000  ...   \n",
       "LINC01409   0.000000       0.0  5.431754  4.811491  6.380960  ...   \n",
       "LINC01128   0.000000       0.0  3.295059  3.752775  4.724329  ...   \n",
       "LINC00115   0.550306       0.0  1.053464  0.000000  0.446060  ...   \n",
       "\n",
       "            Day6_C__283  Day6_C__284  Day6_C__285  Day6_C__286  Day6_C__287  \\\n",
       "AL627309.1          0.0          0.0          0.0     0.000000     0.000000   \n",
       "AL627309.5          0.0          0.0          0.0     0.000000     0.000000   \n",
       "LINC01409           0.0          0.0          0.0     0.236782     0.098364   \n",
       "LINC01128           0.0          0.0          0.0     0.514160     0.187913   \n",
       "LINC00115           0.0          0.0          0.0     0.000000     0.050391   \n",
       "\n",
       "            Day6_C__288  Day6_C__289  Day6_C__290  Day6_C__291  Day6_C__292  \n",
       "AL627309.1          0.0     0.000000          0.0          0.0     0.000000  \n",
       "AL627309.5          0.0     0.000000          0.0          0.0     0.000000  \n",
       "LINC01409           0.0     0.000000          0.0          0.0     0.000000  \n",
       "LINC01128           0.0     2.098073          0.0          0.0     0.000000  \n",
       "LINC00115           0.0     0.084598          0.0          0.0     0.098105  \n",
       "\n",
       "[5 rows x 1382 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "\"\"\"Read the LogNormalised.h5ad file from above and apply the aggregation function to create aggregated data.\n",
    "\"\"\"\n",
    "def aggregateRosa():\n",
    "    adata = scanpy.read(\"data/Rosa/LogNormalised.h5ad\")\n",
    "    celltypes = open(\"data/Rosa/celltypes.tsv\").read().split(\"\\n\")\n",
    "    celltypes = pandas.Series([item for item in celltypes if item!=''], index=adata.obs.index)\n",
    "\n",
    "    agg = pseudoBulkAggregate(adata.to_df().transpose(), celltypes)\n",
    "    samples = pandas.DataFrame({'celltype':[item.split(\"__\")[0] for item in agg.columns]}, index=agg.columns)\n",
    "    print(agg.shape)\n",
    "    display(agg.head())\n",
    "    agg.to_csv(\"data/Rosa/Rosa_aggregated.tsv\", sep=\"\\t\")\n",
    "    samples.to_csv(\"data/Rosa/Rosa_aggregated_samples.tsv\", sep=\"\\t\")\n",
    "    \n",
    "aggregateRosa()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_description</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000283891</th>\n",
       "      <td>MIR628</td>\n",
       "      <td>microRNA 628 [Source:HGNC Symbol;Acc:HGNC:32884]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000251931</th>\n",
       "      <td>RNU6-871P</td>\n",
       "      <td>RNA, U6 small nuclear 871, pseudogene [Source:...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000207766</th>\n",
       "      <td>MIR626</td>\n",
       "      <td>microRNA 626 [Source:HGNC Symbol;Acc:HGNC:32882]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000275323</th>\n",
       "      <td>AC012314.7</td>\n",
       "      <td>40S ribosomal protein S9  [Source:UniProtKB/Sw...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000276678</th>\n",
       "      <td>GHRLOS</td>\n",
       "      <td>Ghrelin opposite strand RNA conserved region [...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  gene_name                                   gene_description\n",
       "gene_id                                                                       \n",
       "ENSG00000283891      MIR628   microRNA 628 [Source:HGNC Symbol;Acc:HGNC:32884]\n",
       "ENSG00000251931   RNU6-871P  RNA, U6 small nuclear 871, pseudogene [Source:...\n",
       "ENSG00000207766      MIR626   microRNA 626 [Source:HGNC Symbol;Acc:HGNC:32882]\n",
       "ENSG00000275323  AC012314.7  40S ribosomal protein S9  [Source:UniProtKB/Sw...\n",
       "ENSG00000276678      GHRLOS  Ghrelin opposite strand RNA conserved region [..."
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(29880, 1382)\n",
      "(26649, 1382)\n"
     ]
    }
   ],
   "source": [
    "# Rosa data comes with gene symbols in the expression matrix - change these to gene ids.\n",
    "# We downloaded gene id to gene symbol mapping from Ensembl Biomart and named the file EnsemblGenes_HomoSapiens_v91.txt.\n",
    "def useEnsemblGeneIds():\n",
    "    ensembl = pandas.read_csv(\"data/Ensembl/EnsemblGenes_HomoSapiens_v91.txt\", sep=\"\\t\", index_col=0)\n",
    "    display(ensembl.head())\n",
    "    geneIdsFromSymbol = {}\n",
    "    for geneId,row in ensembl.iterrows():\n",
    "        if row['gene_name'] not in geneIdsFromSymbol: geneIdsFromSymbol[row['gene_name']] = []\n",
    "        geneIdsFromSymbol[row['gene_name']].append(geneId)\n",
    "    \n",
    "    df = pandas.read_csv(\"data/Rosa/Rosa_aggregated.tsv\", sep=\"\\t\", index_col=0)\n",
    "    print(df.shape)\n",
    "    rows,geneIds = [],[]\n",
    "    for geneSymbol,row in df.iterrows():\n",
    "        for geneId in geneIdsFromSymbol.get(geneSymbol, []):\n",
    "            rows.append(row.tolist())\n",
    "            geneIds.append(geneId)\n",
    "    df = pandas.DataFrame(rows, index=geneIds, columns=df.columns)\n",
    "    print(df.shape)\n",
    "    df.to_csv(\"data/Rosa/Rosa_aggregated_EnsemblId.tsv\", sep=\"\\t\")\n",
    "    \n",
    "useEnsemblGeneIds()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1382, 1)\n",
      "(391, 1) (26649, 391)\n"
     ]
    }
   ],
   "source": [
    "# Aggregated data frame above was still too large for upload to Stemformatics. Reduce by selecting some cell types.\n",
    "# These were the final files used to project onto the Stemformatics Myeloid Atlas and the results are shown in the paper.\n",
    "def createSmallerAggregatedData():\n",
    "    samples = pandas.read_csv(\"data/Rosa/Rosa_aggregated_samples.tsv\", sep=\"\\t\", index_col=0)\n",
    "    print(samples.shape)\n",
    "    celltypes = ['DC1','DC2','PDC','Day3','Day6','Day9','HEF']\n",
    "    samples = samples.loc[[item for item in samples.index if item.split(\"_\")[0] in celltypes and item.split(\"_\")[1]=='B'\\\n",
    "                          or item.startswith('Day9_DP_B')]]\n",
    "    df = pandas.read_csv(\"data/Rosa/Rosa_aggregated_EnsemblId.tsv\", sep=\"\\t\", index_col=0)\n",
    "    df = df[samples.index]\n",
    "    print(samples.shape, df.shape)\n",
    "    df.to_csv(\"data/Rosa/Rosa_aggregated_EnsemblId_subset.tsv\", sep=\"\\t\")\n",
    "    samples.to_csv(\"data/Rosa/Rosa_aggregated_samples_subset.tsv\", sep=\"\\t\")\n",
    "    \n",
    "createSmallerAggregatedData()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "     active environment : scAnalysis\n",
      "    active env location : C:\\Users\\jarnyc\\Anaconda3\\envs\\scAnalysis\n",
      "            shell level : 2\n",
      "       user config file : C:\\Users\\jarnyc\\.condarc\n",
      " populated config files : C:\\Users\\jarnyc\\.condarc\n",
      "          conda version : 23.3.1\n",
      "    conda-build version : 3.24.0\n",
      "         python version : 3.8.16.final.0\n",
      "       virtual packages : __archspec=1=x86_64\n",
      "                          __win=0=0\n",
      "       base environment : C:\\Users\\jarnyc\\Anaconda3  (writable)\n",
      "      conda av data dir : C:\\Users\\jarnyc\\Anaconda3\\etc\\conda\n",
      "  conda av metadata url : None\n",
      "           channel URLs : https://repo.anaconda.com/pkgs/main/win-64\n",
      "                          https://repo.anaconda.com/pkgs/main/noarch\n",
      "                          https://repo.anaconda.com/pkgs/r/win-64\n",
      "                          https://repo.anaconda.com/pkgs/r/noarch\n",
      "                          https://repo.anaconda.com/pkgs/msys2/win-64\n",
      "                          https://repo.anaconda.com/pkgs/msys2/noarch\n",
      "          package cache : C:\\Users\\jarnyc\\Anaconda3\\pkgs\n",
      "                          C:\\Users\\jarnyc\\.conda\\pkgs\n",
      "                          C:\\Users\\jarnyc\\AppData\\Local\\conda\\conda\\pkgs\n",
      "       envs directories : C:\\Users\\jarnyc\\Anaconda3\\envs\n",
      "                          C:\\Users\\jarnyc\\.conda\\envs\n",
      "                          C:\\Users\\jarnyc\\AppData\\Local\\conda\\conda\\envs\n",
      "               platform : win-64\n",
      "             user-agent : conda/23.3.1 requests/2.28.1 CPython/3.8.16 Windows/10 Windows/10.0.19044\n",
      "          administrator : False\n",
      "             netrc file : None\n",
      "           offline mode : False\n",
      "\n",
      "pandas version 1.4.3\n",
      "scanpy version 1.9.1\n"
     ]
    }
   ],
   "source": [
    "# Show conda environment which was used to run this notebook, as well as versions of key packages\n",
    "!conda info\n",
    "print(\"pandas version\", pandas.__version__)\n",
    "print(\"scanpy version\", scanpy.__version__)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
